doc/monads/monads.lean

```diff
@@ -0,0 +1,307 @@
+  pure {α : Type u} : α → f α
+  map : {α β : Type u} → (α → β) → f α → f β
+  seq : {α β : Type u} → f (α → β) → (Unit → f α) → f β
+  bind : {α β : Type u} → f α → (α → f β) → f β
+  ...
+```
+
+Notice `Monad` also contains `pure` it must also have a "default" way to wrap a value in the
+structure.
+
+The `bind` operator is similar to the applicative `seq` operator in that it chains two operations,
+with one of them being function related.
+
+Notice that `bind`, `seq` and `map` all take a function of some kind, and then a structure wrapped
+over a type `α`, namely `f α`, and then produce a structure wrapping a type `β`, namely `f β`. They
+just vary in what the function looks like. For the functor, the function is a normal pure function
+because it has the type `(α → β)`. For applicatives, the function is still pure, but wrapped in the
+structure `f (α → β)`. Now with monads, the function argument takes a "pure" input `α` but produces
+an output in the structure, `(α → f β)`
+
+## Basic Monad Example
+
+Just as `Option` is a functor and an applicative functor, it is also a monad!
+Let's start with how `Option` implements the Monad type class.
+
+-/
+instance : Monad Option where
+  pure := Option.some
+  bind := Option.bind
+/-!
+
+where:
+
+```lean,ignore
+def Option.bind : Option α → (α → Option β) → Option β
+  | none,   _ => none
+  | some a, f => f a
+```
+
+> **Side note**: this function definition is using a special shorthand syntax in Lean where the `:=
+match a, b with` code can be collapsed away. To make this more clear consider the following simpler
+example, where `Option.bind` is using the second form like `bar`:
+
+-/
+def foo (x : Option Nat) (y: Nat) : Option Nat :=
+  match x, y with
+  | none, _ => none
+  | some x, y => some (x + y)
+
+def bar : Option Nat → Nat → Option Nat
+  | none, _ => none
+  | some x, y => some (x + y)
+
+#eval foo (some 1) 2  -- some 3
+#eval bar (some 1) 2  -- some 3
+/-!
+What is important is that `Option.bind` is using a `match` statement to unwrap the input value
+`Option α`, if it is `none` then it does nothing and returns `none`, if it has a value of type `α`
+then it applies the function in the second argument `(α → Option β)` to this value, which is
+the expression `f a` that you see in the line `  | some a, f => f a` above.  The function
+returns a result of type `Option β` which then becomes the return value for `bind`.  So there
+is no structure wrapping required on the return value since the input function already did that.
+
+But let's bring in the definition of a monad. What does it mean to describe `Option` as a
+computational context?
+
+The `Option` monad encapsulates the context of failure. Essentially, the `Option` monad lets us
+abort a series of operations whenever one of them fails. This allows future operations to assume
+that all previous operations have succeeded. Here's some code to motivate this idea:
+
+-/
+def optionFunc1 : String -> Option Nat
+| "" => none
```

Comment:
Please indent the `|`s.

Reply:
fixed in https://github.com/leanprover/lean4/pull/1505/commits/e53237c5a9afcfcfd2536283350598f393081ab3